### ISSUE TITLE
Consolidate the various approaches we have to projecting from the kernel heap

### ISSUE THREAD

**Issue by michaelmcinerney** [seL4/l4v#841]
We currently have at least two approaches to performing heap projections, namely the `pred_map` style, and the `|>` (`opt_map`) style. 

The `pred_map` style uses locales together with collections of simplification rules to define and reason with the heap projections, whereas `|>` is a much more lightweight approach. Given the difficulty of extending the `pred_map` approach to chains of heaps which have not already been defined, we do prefer the `|>` approach, which has adequate and simple simplification rules, as well. We face problems especially with lemmas and goals that contain both approaches.

We should consolidate on the `|>` approach. This in particular would require rewriting all the invariants for MCS given in `DetSchedInvs_AI.thy`. 

More ambitious projects include 
 - making greater use of these heap projections in the abstract specification, and
 - phrasing invariants which use `obj_at/obj_at'` in terms of `|>` and `|<` (`opt_pred`).

**Comment by Xaphiosis** [comment]
I support this motion, and agree that it is the most sensible first step. If we got this, would it be clear that the `hoare_lift_Pf` blocks will be able to go away? (you didn't mention those, and they were the inspiration for my pushing for an issue)

**Comment by michaelmcinerney** [comment]
> I support this motion, and agree that it is the most sensible first step. If we got this, would it be clear that the `hoare_lift_Pf` blocks will be able to go away? (you didn't mention those, and they were the inspiration for my pushing for an issue)

Yes, those `hoare_lift_Pf` blocks only occur in the `misc` lemmas. We might need a bit of finesse in how we handle the replacements for these `misc` lemmas: `crunch` is a great tool, but we may want a way to give `crunch` a large collection of heaps, and it would be nice if it were easier to spot which of the two dozen or so properties we're crunching is leading to a failure.

**Comment by Xaphiosis** [comment]
Oh, I thought for some reason it also happened in non-misc lemmas. They certainly seemed to make a more common appearance.

Also, looking through issues I found one that's semi-related to this: https://github.com/seL4/l4v/issues/633

**Comment by michaelmcinerney** [comment]
> Oh, I thought for some reason it also happened in non-misc lemmas. They certainly seemed to make a more common appearance.

I suppose they also do occur in lemmas with postconditions of `valid_sched_pred_strong` and `valid_sched_pred`, but those are really just abbreviations for a bunch of heaps too. I guess the name "misc" just means "almost like `valid_sched_pred_strong` or something, but not quite, so it's miscellaneous".

**Comment by Xaphiosis** [comment]
Ewww. Definitely something we should keep in our sights to improve upon then. I hope the memory of our discussion about lifting rules will remain in your head in case it's ever useful, because until the pred_map/opt_map consolidation my ideas don't really help.

**Comment by lsf37** [comment]
Issue #854 is slightly related to this in the sense that it wants to make more use of projections in ASpec.
